tests/test_mpx_inner.py

```diff
@@ -0,0 +1,230 @@
+        Positional offsets. The implementation is testing by explicitly
+        computing cross covariance differences at these offsets, then comparing
+        to the outcome of the same thing, computed using difference equations.
+
+    Returns
+    -------
+    df, dg : array_like, array_like
+    
+    Test diff_equations through the python interface compute_dfdg
+    
+    This verifies that for four windows of a time series, 
+
+    if U = T[i : i + w] - mu[i],
+        V = T[j : j + w] - mu[j],
+        W = T[i + 1 : i + w + 1] - mu[i + 1],
+        X = T[j + 1 : j + w + 1] - mu[j + 1]
+        0 <= i,j <= len(T) - w
+
+    then
+
+       <W, X>  - <U, V>  == df[i + 1] * dg[j + 1] + df[j + 1] * dg[i + 1]
+
+    with df, dg computed by the corresponding function.
+
+    """
+
+    sseqct = mu.shape[0]
+    w = ts.shape[0] - sseqct + 1
+    df = np.empty(sseqct, dtype='d')
+    dg = np.empty(sseqct, dtype='d')
+
+    cympx_inner.compute_diff_eqns(df, dg, ts, mu)
+    
+    for offset in test_offsets:
+        cmtct = sseqct - offset
+        cov = np.empty(cmtct, dtype='d')
+    
+        for i in range(cmtct):
+            cov[i] = np.dot(ts[i:i+w] - mu[i], ts[i+offset:i+offset+w] - mu[i+offset])
+    
+        desired = np.diff(cov)
+        actual = np.empty(cmtct-1, dtype='d')
+
+        for i in range(cov.shape[0] - 1):
+            # the offset by 1 covers that we're indexing differences only here
+            actual[i] = df[i+1] * dg[i+offset+1] + df[i+offset+1] * dg[i+1]
+  
+            # We can use a tighter tolerance if the parts that aren't being explicitly tested here
+            # are computed using an arbitrary precision library.
+        
+        np.testing.assert_allclose(actual, desired, rtol=1e-8)
+
+    return df, dg
+
+
+def cross_cov_with_verif(ts, mu, cmpseq):
+    """
+    Tests unscaled cross covariance between a time series and the first window of the time series.
+    Also returns the computed result to be used in downstream tests.
+
+    Parameters
+    ----------
+
+    ts : array_like
+        The input time series.
+
+    cmpseq: array_like
+        Mean centered comparison sequence, defaults to the first in the time series.
+
+    Outputs
+    -------
+    cov: array_like
+        Unscaled co-moment comparing ts[:w] - mu[0] to each mean centered window ts[:w] - mu[0] ..... ts[-w:] - mu[-w].
+
+    """
+    w = cmpseq.shape[0]
+    sseqct = ts.shape[0] - w + 1
+    assert(mu.shape[0] == sseqct)
+    np.testing.assert_almost_equal(np.sum(cmpseq), 0)
+    actual = np.empty(sseqct)
+    cympx_inner.compute_cross_cov(actual, ts, mu, cmpseq)
+    desired = np.empty(sseqct)
+    for i in range(sseqct):
+        desired[i] = np.dot(ts[i : i + w] - mu[i], cmpseq)  
+    # If I set tolerance too tight here, it fails.
+    # Testing against numpy results is a little tenuous. 
+    # We should use a high precision library to aid in verification. 
+    np.testing.assert_allclose(actual, desired, rtol=1e-10)
```

Comment:
What do you suggest for testing precision?

Reply:
Ah I forgot I left that note in there. 

A high precision library like GMP would be good. I have been searching for one with up to date Python bindings. The 2 to 3 transition killed off a lot of these.

Reply:
I feel like this could be an improvement later on. A lot of libraries tend to use numpy for testing as we do. What do you think?

Reply:
That is reasonable. I'll remove the comment from the source. 

As I mentioned elsewhere, a robust test that doesn't rely on arbitrary precision libraries would be to write back the final co-moment value to cov for each sequence of updates, then check it against an explicitly computed co-moment. 

This would tell you more about whether calculations diverge, regardless of nearest neighbor correctness, and it's within Numpy's test capabilities.